app/src/main/java/com/android/sample/model/map/LocationViewModel.kt

```diff
@@ -55,4 +60,51 @@ constructor(
       // Handle the permission denial as appropriate
     }
   }
+
+  /**
+   * Calculate the distance between the current location and the given activity location.
+   *
+   * @param activityLocation The location of the activity.
+   * @return The distance in kilometers.
+   */
+  fun getDistanceFromCurrentLocation(activityLocation: Location?): Float? {
+    if (activityLocation == null) return null
+    val currentLocation = _currentLocation.value ?: return null
+    return calculateDistance(
+            currentLocation.latitude,
+            currentLocation.longitude,
+            activityLocation.latitude,
+            activityLocation.longitude)
+        .toFloat()
+  }
+
+  /**
+   * Calculate the distance between two locations using the Haversine formula.
+   *
+   * @param lat1 Latitude of the first location.
+   * @param lon1 Longitude of the first location.
+   * @param lat2 Latitude of the second location.
+   * @param lon2 Longitude of the second location.
+   * @return The distance in kilometers.
+   */
+  private fun calculateDistance(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
+    val earthRadius = 6371.0 // Radius of the Earth in kilometers
+
+    // Convert latitude and longitude from degrees to radians
+    val lat1Rad = Math.toRadians(lat1)
+    val lon1Rad = Math.toRadians(lon1)
+    val lat2Rad = Math.toRadians(lat2)
+    val lon2Rad = Math.toRadians(lon2)
+
+    // Haversine formula
+    val dLat = lat2Rad - lat1Rad
+    val dLon = lon2Rad - lon1Rad
+
+    val a = sin(dLat / 2).pow(2) + cos(lat1Rad) * cos(lat2Rad) * sin(dLon / 2).pow(2)
+
+    val c = 2 * atan2(sqrt(a), sqrt(1 - a))
+
+    // Distance in kilometers
+    return earthRadius * c
+  }
```

Comment:
Thanks for providing the comments documentations. The functions speak for themselves. Weird how you couldn't use android.location.Location distanceTo function! But I guess it is okay, as you used an official calculation function. The math implementation seems correct. I just compared it with the original formula.